contracts/NFTStorefrontV2.cdc

```diff
@@ -0,0 +1,696 @@
+        access(contract) let nftProviderCapability: Capability<&{NonFungibleToken.Provider, NonFungibleToken.CollectionPublic}>
+
+        // Optional list of marketplaces capabilities, Used for incentivising 
+        // them if purchase gets facilitated by them.
+        access(contract) let marketplacesCap: [Capability<&{FungibleToken.Receiver}>]?
+
+        // borrowNFT
+        // This will assert in the same way as the NFT standard borrowNFT()
+        // if the NFT is absent, for example if it has been sold via another listing.
+        //
+        pub fun borrowNFT(): &NonFungibleToken.NFT {
+            let ref = self.nftProviderCapability.borrow()!.borrowNFT(id: self.getDetails().nftID)
+            assert(ref.isInstance(self.getDetails().nftType), message: "NFT has wrong type")
+            assert(ref.id == self.getDetails().nftID, message: "NFT has wrong ID")
+            return ref as &NonFungibleToken.NFT
+        }
+
+        // getDetails
+        // Get the details of the current state of the Listing as a struct.
+        // This avoids having more public variables and getter methods for them, and plays
+        // nicely with scripts (which cannot return resources). 
+        //
+        pub fun getDetails(): ListingDetails {
+            return self.details
+        }
+
+        // purchase
+        // Purchase the listing, buying the token.
+        // This pays the beneficiaries and commission to the facilitator and returns extra token to the buyer.
+        // This also cleans up duplicate listings for the item being purchased.
+        pub fun purchase(
+            payment: @FungibleToken.Vault, 
+            commissionRecipient: Capability<&{FungibleToken.Receiver}>,
+        ): @NonFungibleToken.NFT {
+
+            pre {
+                self.details.purchased == false: "listing has already been purchased"
+                payment.isInstance(self.details.salePaymentVaultType): "payment vault is not requested fungible token"
+                payment.balance == self.details.salePrice: "payment vault does not contain requested price"
+                self.details.expiry > UInt64(getCurrentBlock().timestamp): "Listing is expired"
+                self.owner != nil : "Resource doesn't have the assigned owner"
+            }
+            // Access the StoreFrontManager resource reference to remove the duplicate listings if purchase would happen successfully.
+            let storeFrontPublicRef = self.owner!.getCapability<&NFTStorefrontV2.Storefront{NFTStorefrontV2.StorefrontPublic}>(NFTStorefrontV2.StorefrontPublicPath)
+                                        .borrow() ?? panic("Unable to borrow the storeFrontManager resource")
+            // Make sure the listing cannot be purchased again.
+            self.details.setToPurchased() 
+            
+            if (self.details.commissionAmount > 0.0) {
+                if self.marketplacesCap != nil {
+                    var doesCapabilityAllowed = false
```

Comment:
```
                    var isCommissionRecipientValid = false
```

Reply:
nice - glad to see the validation to ensure the commission recipient has to be listed as one of the marketplace addresses...